### PR TITLE
TDL-18010: Add custom exception handling

### PR DESCRIPTION
**Pull request by savan-chovatiya**
# Description of change
- Added custom exceptions and custom messages for status code mentioned in TwitterAds API.

# Manual QA steps
 - Provided wrong credentials in the config and verified that tap is failing with custom exception and proper message..

 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_twitter_ads/sync.py [hunk 1]
```diff
@@ -120,10 +121,12 @@ def get_async_data(report_name, client, url):
...
             client, 'get', resource.path, domain=domain, raw_body=True, stream=True).perform()
         response_body = response.body
-    except Error as err:
-        # see twitter_ads.error for more details
-        LOGGER.error('Report: {} - ERROR: {}'.format(report_name, err.details))
-        raise err
+
+    except Exception as e:
+        LOGGER.error('Report: {} - ERROR: {}'.format(report_name, e.details))
+        # see tap-twitter-ads.client for more details
```
Comment (umangagrawal-crest) [inline tap_twitter_ads/sync.py:127]:
```suggestion
        # see tap-twitter-ads.client for more details
        LOGGER.error('Report: {} - ERROR: {}'.format(report_name, e.details))
        
```

**Review by KrisPersonal — APPROVED** [review]

File: tap_twitter_ads/client.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    },
+    429: {
+        "raise_exception": TwitterAdsClient429Error,
+        "message": "API rate limit exceeded, please retry after some time."
+    },
+    500: {
+        "raise_exception": TwitterAdsInternalServerError,
+        "message": "Internal error."
+    },
+
```
Comment (dbshah1212) [inline tap_twitter_ads/client.py:69]: Remove extra line

File: tap_twitter_ads/sync.py (hunk 1 above)
Comment (savan-chovatiya) [reply tap_twitter_ads/sync.py]: Updated

File: tap_twitter_ads/client.py (hunk 2 above)
Comment (savan-chovatiya) [reply tap_twitter_ads/client.py]: Removed extra line.

**Review by dbshah1212 — APPROVED** [review]

**Review by namrata270998 — APPROVED** [review]

File: tests/unittests/test_error_handling.py [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+        """
+            Test case to verify 400 error message from response
+        """
+        
+        with self.assertRaises(client.TwitterAdsBadRequestError) as e:
+            self.test_stream.get_resource("test", self.client_obj, "path")
+
+        self.assertEqual(str(e.exception), "HTTP-error-code: 400, Message: The request is missing or has a bad parameter.")
+    
+    @mock.patch("tap_twitter_ads.streams.Cursor", side_effect=Mockresponse(400, [{"message":"This mesaage from response 400"}]))
```
Comment (RushT007) [inline tests/unittests/test_error_handling.py:42]: Spelling mistake `mesaage` in all unittests, please correct.

**Review by RushT007 — CHANGES_REQUESTED** [review]

File: tests/unittests/test_error_handling.py (hunk 3 above)
Comment (savan-chovatiya) [reply tests/unittests/test_error_handling.py]: Resolve spell mistakes in unit tests.

**Review by kspeer825 — APPROVED** [review]

**Review by RushT007 — APPROVED** [review]

**Pull request merged**